File: go/jpl/params.go
package jpl

import "github.com/jplorg/jpl/go/definition"

type JPLInstructionParams struct {
	After          string
	ArgNames       []string
	Assignment     *JPLAssignment
	Catch          JPLFunc
	Comparisons    []JPLComparison
	Else           JPLFunc
	Fields         []JPLField
	Ifs            []JPLIfThen
	Interpolations []JPLInterpolation
	Name           string
	Operations     []JPLOperation
	Pipe           JPLFunc
	Pipes          []JPLFunc
	Selectors      []JPLSelector
	Try            JPLFunc
	Value          any
	String         string
	Number         float64
}

type JPLComparison struct {
	OP     definition.JPLOPC
	Params JPLComparisonParams
}

type JPLComparisonParams struct {
	By JPLFunc
}

type JPLSelector struct {
	OP     definition.JPLOPA
	Params JPLSelectorParams
}

type JPLSelectorParams struct {
	Args     []JPLFunc
	Bound    bool
	From     JPLFunc
	Optional bool
	Pipe     JPLFunc
	To       JPLFunc
}

type JPLOperation struct {
	OP     definition.JPLOPM
	Params JPLOperationParams
}

type JPLOperationParams struct {
	By JPLFunc
}

type JPLIfThen struct {
	If   JPLFunc
	Then JPLFunc
}

type JPLAssignment struct {
	OP     definition.JPLOPU
	Params JPLAssignmentParams
}

type JPLAssignmentParams struct {
	Pipe JPLFunc
}

type JPLField struct {
	Key      JPLFunc
	Value    JPLFunc
	Optional bool
}

type JPLInterpolation struct {
	Before string
	Pipe   JPLFunc
}

File: js/lib/program/params.ts
import type {
  JPLFunc,
  OPA_FIELD,
  OPA_FUNCTION,
  OPA_ITER,
  OPA_SLICE,
  OPC_EQUAL,
  OPC_GREATER,
  OPC_GREATEREQUAL,
  OPC_LESS,
  OPC_LESSEQUAL,
  OPC_UNEQUAL,
  OPM_ADDITION,
  OPM_DIVISION,
  OPM_MULTIPLICATION,
  OPM_REMAINDER,
  OPM_SUBTRACTION,
  OPU_ADDITION,
  OPU_DIVISION,
  OPU_MULTIPLICATION,
  OPU_NULL_COALESCENCE,
  OPU_REMAINDER,
  OPU_SET,
  OPU_SUBTRACTION,
  OPU_UPDATE,
} from '@/library';

export type UserJPLInstructionParams = {
  after?: string;
  argNames?: string[];
  assignment?: UserJPLAssignment;
  catch?: JPLFunc;
  comparisons?: UserJPLComparison[];
  else?: JPLFunc;
  fields?: UserJPLField[];
  ifs?: UserJPLIfThen[];
  interpolations?: UserJPLInterpolation[];
  name?: string;
  number?: number;
  operations?: UserJPLOperation[];
  pipe?: JPLFunc;
  pipes?: JPLFunc[];
  selectors?: UserJPLSelector[];
  string?: string;
  try?: JPLFunc;
  value?: unknown;
};

export type UserJPLComparison = {
  op:
    | typeof OPC_EQUAL
    | typeof OPC_GREATER
    | typeof OPC_GREATEREQUAL
    | typeof OPC_LESS
    | typeof OPC_LESSEQUAL
    | typeof OPC_UNEQUAL;
  params: UserJPLComparisonParams;
};

export type UserJPLComparisonParams = { by: JPLFunc };

export type UserJPLSelector = {
  op:
    | typeof OPA_FIELD
    | typeof OPA_FUNCTION
    | typeof OPA_ITER
    | typeof OPA_SLICE;
  params: UserJPLSelectorParams;
};

export type UserJPLSelectorParams = {
  args?: JPLFunc[];
  bound?: boolean;
  from?: JPLFunc;
  optional?: boolean;
  pipe?: JPLFunc;
  to?: JPLFunc;
};

export type UserJPLOperation = {
  op:
    | typeof OPM_ADDITION
    | typeof OPM_DIVISION
    | typeof OPM_MULTIPLICATION
    | typeof OPM_REMAINDER
    | typeof OPM_SUBTRACTION;
  params: UserJPLOperationParams;
};

export type UserJPLOperationParams = { by: JPLFunc };

export type UserJPLIfThen = { if: JPLFunc; then: JPLFunc };

export type UserJPLAssignment = {
  op:
    | typeof OPU_ADDITION
    | typeof OPU_DIVISION
    | typeof OPU_MULTIPLICATION
    | typeof OPU_NULL_COALESCENCE
    | typeof OPU_REMAINDER
    | typeof OPU_SET
    | typeof OPU_SUBTRACTION
    | typeof OPU_UPDATE;
  params: UserJPLAssignmentParams;
};

export type UserJPLAssignmentParams = { pipe: JPLFunc };

export type UserJPLField = { key: JPLFunc; value: JPLFunc; optional: boolean };

export type UserJPLInterpolation = { before: string; pipe: JPLFunc };
